rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+
+# RFC 005: Soft Chain Upgrades
+
+## Changelog
+
+- 2020-11-27: Add appendix to distinguish between soft and hard upgrades
+- 2020-11-23: Added chain migration proposal for comparison
+- 2020-11-16: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+The current method for upgrading a network across block protocol breaking
+changes is to hard fork the chain. All nodes must coordinate a halt height/
+time, where application state can be exported across and an entirely new chain
+(as far as Tendermint is concerned) is created (see [cosmoshub-3
+upgrade](https://blog.cosmos.network/cosmos-hub-3-upgrade-announcement-39c9da941aee)).
+This means that each block protocol version maps to the entire life of a single
+chain. The result is that networks either forego new features for long periods
+of time or have to deal with the pains of upgrading such as the coordination
+effort, state migration, intra-chain security challenges and most importantly,
+the loss of transaction history.
+
+The scope of this RFC is to address the need for Tendermint to have better
+flexibility in improving itself whilst offering greater ease of use to the
+networks running on Tendermint. This is done by introducing a division in the
+way a network can upgrade: either soft or hard. More information on the
+distinction as well as general upgrade terminology can be found in Appendix A.
+There may still remain the need for hard upgrades but this will be focused
+[elsewhere](https://github.com/tendermint/tendermint/issues/5595)). Instead,
+this RFC will describe what supporting soft upgrades would entail and explore
+two different methods: **Multi Data Structure Support** or
+**Chain Migrations**.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+A soft upgrade, in the simplest terms, is a change to Tendermint's binary that
+would not impede with the use of prior data / existing chains.
+
+The implementation to support this would ideally need to be done prior to the
+1.0 release.
+
+### Stakeholders
+
+Let's first set out to define the different needs that the various stakeholders
+of the Tendermint software have in the context of upgrading:
+
+- Application Developers: want to be able to get as much of the benefits of
+  new features (like state sync and light client) with minimal work or loss to
+  their current infrastructure. Ideally, Tendermint is also somewhat
+  accommodating to application upgrades.
+
+- Node Operators: Primarily they want safety and reliability. This means minimal
+  down-time and intuitive / easy UX when it comes to upgrading.
+
+- Wallets, Block Explorers and other clients: data retrievability which means
+  ensuring that helpful/informative data is always available and that minimal
+  infrastructure is needed to support serving data across the entire lifespan of
+  the chain.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
```

Comment:
Do you think it would be helpful to add more context on this? I didn't personally come up with the protocol versions but the abstraction does make sense in terms of separate parts of the code.

Another way that I thought this could be done is to look at it from the stakeholders angle. We kind of have three stakeholders: the application developers, the node operators, and the "end users" which could be wallets and block explorers. A nice way to version things would be to indicate whether the changes require an involvement from that stakeholder. So ABCI and App version correlates to application developers, RPC version (the one I proposed) correlates with external clients. I'm not sure where the node operators fit but they would need to make changes to their binaries to support different block versions (and potentially p2p versions).